styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java

```diff
@@ -127,38 +135,110 @@
+      final Set<String> activeSecrets = pods.getItems().stream()
+          .flatMap(pod -> pod.getSpec().getVolumes().stream())
+          .map(volume -> volume.getSecret().getSecretName())
+          .collect(Collectors.toSet());
+
+      // Enumerate all service account secrets that are not in active use
+      final SecretList secrets = client.secrets().list();
+      final List<Secret> inactiveServiceAccountSecrets = secrets.getItems().stream()
+          .filter(secret -> secret.getMetadata().getName().startsWith(STYX_WORKFLOW_SA_SECRET_NAME))
+          .filter(secret -> !activeSecrets.contains(secret.getMetadata().getName()))
+          .collect(Collectors.toList());
+
+      // Delete keys and secrets for all inactive service accounts and let them be recreated by future executions
+      for (Secret secret : inactiveServiceAccountSecrets) {
+        final String name = secret.getMetadata().getName();
+        final String serviceAcountEmail = serviceAccountEmail(secret);
+
+        try {
+          final Map<String, String> annotations = secret.getMetadata().getAnnotations();
+          final String jsonKeyName = annotations.get(STYX_WORKFLOW_SA_JSON_KEY_NAME_ANNOTATION);
+          final String p12KeyName = annotations.get(STYX_WORKFLOW_SA_P12_KEY_NAME_ANNOTATION);
+
+          LOG.info("[AUDIT] Deleting service account key: {}", jsonKeyName);
+          tryDeleteServiceAccountKey(jsonKeyName);
+
+          LOG.info("[AUDIT] Deleting service account key: {}", p12KeyName);
+          tryDeleteServiceAccountKey(p12KeyName);
+
+          LOG.info("[AUDIT] Deleting service account {} secret {}", serviceAcountEmail, name);
+          client.secrets().delete(secret);
+        } catch (IOException | KubernetesClientException e) {
+          LOG.warn("[AUDIT] Failed to delete service account {} keys and/or secret {}",
+              serviceAcountEmail, name);
+        }
+      }
+    }
+  }
+
+  /**
+   * Try to delete a service account key, giving up with a warning if permission was denied.
+   * @param keyName The fully qualified name of the key to delete.
+   */
+  private void tryDeleteServiceAccountKey(String keyName) throws IOException {
     try {
-      final Pod pod = client.pods().create(createPod(workflowInstance, runSpec));
-      return pod.getMetadata().getName();
-    } catch (KubernetesClientException kce) {
-      throw new IOException("Failed to create Kubernetes pod", kce);
+      serviceAccountKeyManager.deleteKey(keyName);
+    } catch (GoogleJsonResponseException e) {
+      if (GcpUtil.isPermissionDenied(e)) {
+        LOG.warn("[AUDIT] Permission denied when trying to delete unused service account key {}");
+      } else {
+        throw e;
+      }
     }
   }
 
-  private void ensureSecrets(WorkflowInstance workflowInstance, RunSpec runSpec) throws IOException {
-    ensureServiceAccountKeySecret(workflowInstance, runSpec);
+  private String serviceAccountEmail(Secret secret) {
```

Comment:
It looks we never call it `...Email` in other places, so maybe just `serviceAccount` or `getServiceAccount`? Or maybe just inline if splitting `cleanup`.